### PR TITLE
ddl: implement `ALTER DATABASE` to alter charset/collation

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 

Add support for `ALTER DATABASE` syntax

Related issue: pingcap/tidb#4641
Related parser PR: https://github.com/pingcap/parser/pull/318

### What is changed and how it works?
A new DDL job is added to change charset/collate for a database.

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

**Comment by bb7133** [comment]
PTAL @crazycs520 @lonng @winkyao

File: go.sum1 [hunk 1]
```diff
@@ -0,0 +1,179 @@
+github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=
```
Comment (jackysp) [inline go.sum1]: Why we need this file?

**Comment by crazycs520** [comment]
@bb7133  I got some problem, but maybe we can fix this in other PR:
In this PR
```sql
mysql root@localhost:(none)> create database test charset utf8;
Query OK, 0 rows affected
Time: 0.005s
mysql root@localhost:(none)> alter database test charset="utf8"
(1105, u'unsupported modify collate from  to utf8_bin')
mysql root@localhost:(none)> drop database if exists test
Query OK, 0 rows affected
Time: 0.010s
mysql root@localhost:(none)> create database test charset UTF8MB4;
Query OK, 0 rows affected
Time: 0.005s
mysql root@localhost:(none)> alter database test charset="utf8mb4"
(1105, u'unsupported modify collate from  to utf8mb4_bin')
mysql root@localhost:(none)>
```

There is some bug in create a schema.

**Comment by bb7133** [comment]
> There are some bugs in creating a schema.

https://github.com/pingcap/tidb/blob/d241d741b9c8e5c480babf690fcd1b980c7af615/ddl/ddl_api.go#L2170

Isn't this an expected behavior? For now, we only support changing charset from `utf8` to `utf8mb4`

File: go.sum1 (hunk 1 above)
Comment (bb7133) [reply go.sum1]: It's a mistake, thank you very much!

**Comment by bb7133** [comment]
@jackysp @crazycs520 addressed, PTAL, thank you!

File: ddl/schema.go [hunk 2]
```diff
@@ -109,6 +111,35 @@ func checkSchemaNotExistsFromStore(t *meta.Meta, schemaID int64, dbInfo *model.D
...
+func onModifySchemaCharsetAndCollate(t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	var toCharset, toCollate string
+	if err := job.DecodeArgs(&toCharset, &toCollate); err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	dbInfo, err := checkSchemaExistAndCancelNotExistJob(t, job)
+	if err != nil {
+		if infoschema.ErrDatabaseDropExists.Equal(err) {
```
Comment (winkyao) [inline ddl/schema.go]: This error is weird, I think to call `checkSchemaExistAndCancelNotExistJob` is not appropriate. Maybe you should call `t.GetDatabase(job.SchemaID)` directly.

File: infoschema/builder.go [hunk 3]
```diff
@@ -127,6 +129,30 @@ func (b *Builder) applyCreateSchema(m *meta.Meta, diff *model.SchemaDiff) error
...
+		// When we apply an old schema diff, the database may has been dropped already
+		return ErrDatabaseNotExists.GenWithStackByArgs(
+			fmt.Sprintf("(Schema ID %d)", diff.SchemaID),
+		)
+	}
+	oldInfo, ok := b.is.SchemaByName(di.Name)
+	if !ok {
+		return ErrDatabaseNotExists.GenWithStackByArgs(di.Name.O)
+	}
+	if oldInfo.Charset == di.Charset && oldInfo.Collate == di.Collate {
```
Comment (zimulala) [inline infoschema/builder.go]: If this happens, it is already processed when the DDL job is processed. Doesn't this happen here?

File: infoschema/builder.go [hunk 4]
```diff
@@ -127,6 +129,30 @@ func (b *Builder) applyCreateSchema(m *meta.Meta, diff *model.SchemaDiff) error
 	return nil
 }
 
+func (b *Builder) applyModifySchemaCharsetAndCollate(m *meta.Meta, diff *model.SchemaDiff) error {
+	di, err := m.GetDatabase(diff.SchemaID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if di == nil {
+		// When we apply an old schema diff, the database may has been dropped already
```
Comment (zimulala) [inline infoschema/builder.go]: Ditto. Maybe we can update this comment.

File: ddl/schema.go [hunk 5]
```diff
@@ -109,6 +111,35 @@ func checkSchemaNotExistsFromStore(t *meta.Meta, schemaID int64, dbInfo *model.D
...
+			desc := fmt.Sprintf("(Schema ID %d)", job.ID)
+			return ver, infoschema.ErrDatabaseNotExists.GenWithStackByArgs(desc)
+		}
+		return ver, errors.Trace(err)
+	}
+
```
Comment (zimulala) [inline ddl/schema.go]: I think we can check "if dbInfo.Charset == toCharset && dbInfo.Charset == toCollate {return nil}" here。

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -90,6 +90,63 @@ func (d *ddl) CreateSchema(ctx sessionctx.Context, schema model.CIStr, charsetIn
...
+	if dbInfo.Charset == toCharset && dbInfo.Charset == toCollate {
+		return nil
+	}
+
+	// check current TiDB limitations
+	if err = modifiableCharsetAndCollation(toCharset, toCollate, dbInfo.Charset, dbInfo.Collate); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Do DDL job
```
Comment (zimulala) [inline ddl/ddl_api.go]:
```suggestion
	// Do the DDL job.

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -90,6 +90,63 @@ func (d *ddl) CreateSchema(ctx sessionctx.Context, schema model.CIStr, charsetIn
...
+	is := d.GetInfoSchemaWithInterceptor(ctx)
+	dbInfo, ok := is.SchemaByName(dbName)
+	if !ok {
+		return infoschema.ErrDatabaseNotExists.GenWithStackByArgs(dbName.O)
+	}
+	if dbInfo.Charset == toCharset && dbInfo.Charset == toCollate {
+		return nil
+	}
+
+	// check current TiDB limitations
```
Comment (zimulala) [inline ddl/ddl_api.go]:
```suggestion
	// Check the current TiDB limitations.

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -90,6 +90,63 @@ func (d *ddl) CreateSchema(ctx sessionctx.Context, schema model.CIStr, charsetIn
 	err = d.callHookOnChanged(err)
 	return errors.Trace(err)
 }
+func (d *ddl) AlterSchema(ctx sessionctx.Context, stmt *ast.AlterDatabaseStmt) (err error) {
+	// Resolve target charset and collation from options
```
Comment (zimulala) [inline ddl/ddl_api.go]: Please add the `.` at the end of a line. And please update other comments in this PR.

File: infoschema/builder.go (hunk 3 above)
Comment (bb7133) [reply infoschema/builder.go]: I suppose another DDL job may have changed the charset already, is it possible?

File: ddl/schema.go (hunk 5 above)
Comment (bb7133) [reply ddl/schema.go]: OK, then

File: infoschema/builder.go (hunk 3 above)
Comment (zimulala) [reply infoschema/builder.go]: I don't think so. If another DDL job changed, this DDL job needn't to be handled.

File: ddl/schema.go (hunk 2 above)
Comment (bb7133) [reply ddl/schema.go]: addressed

**Comment by bb7133** [comment]
PTAL @winkyao @zimulala , thanks

File: ddl/schema.go [hunk 9]
```diff
@@ -109,6 +109,36 @@ func checkSchemaNotExistsFromStore(t *meta.Meta, schemaID int64, dbInfo *model.D
...
 
+func onModifySchemaCharsetAndCollate(t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	var toCharset, toCollate string
+	if err := job.DecodeArgs(&toCharset, &toCollate); err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	dbInfo, err := t.GetDatabase(job.SchemaID)
+	if err != nil {
```
Comment (zimulala) [inline ddl/schema.go]: Here we need to add `job.State = model.JobStateCancelled` to make sure the job will be done.

Comment (winkyao) [reply ddl/schema.go]: yes, if we don't cancel the job here, the job will never be finished.

File: ddl/schema.go [hunk 10]
```diff
@@ -109,6 +109,37 @@ func checkSchemaNotExistsFromStore(t *meta.Meta, schemaID int64, dbInfo *model.D
...
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	if dbInfo.Charset == toCharset && dbInfo.Collate == toCollate {
+		job.FinishDBJob(model.JobStateDone, model.StatePublic, ver, dbInfo)
```
Comment (zimulala) [inline ddl/schema.go]: I think replace this line with `job.State = model.JobStateCancelled` is better.
We don't do this schema change, becase other job is done.

Comment (bb7133) [reply ddl/schema.go]: But this will cause an error message, right?

Comment (zimulala) [reply ddl/schema.go]: No. Here we return error is nill.

Comment (bb7133) [reply ddl/schema.go]: addressed, thanks @zimulala

**Comment by bb7133** [comment]
run-test/integration-ddl-test

**Pull request merged**